File: app/src/main/java/com/teambrella/android/ui/teammate/TeammateFragment.java
package com.teambrella.android.ui.teammate;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.ObjectAnimator;
import android.graphics.Rect;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.widget.NestedScrollView;
import android.text.Html;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.google.gson.JsonObject;
import com.squareup.picasso.Picasso;
import com.teambrella.android.R;
import com.teambrella.android.api.TeambrellaModel;
import com.teambrella.android.api.model.json.JsonWrapper;
import com.teambrella.android.api.server.TeambrellaServer;
import com.teambrella.android.api.server.TeambrellaUris;
import com.teambrella.android.image.TeambrellaImageLoader;
import com.teambrella.android.ui.base.ADataFragment;
import com.teambrella.android.ui.base.ADataProgressFragment;
import com.teambrella.android.ui.chat.ChatActivity;
import com.teambrella.android.ui.widget.TeambrellaAvatarsWidgets;
import com.teambrella.android.ui.widget.VoterBar;
import com.teambrella.android.util.AmountCurrencyUtil;
import com.teambrella.android.util.ConnectivityUtils;
import com.teambrella.android.util.TeambrellaDateUtils;

import io.reactivex.Notification;
import io.reactivex.Observable;
import jp.wasabeef.picasso.transformations.CropCircleTransformation;

/**
 * Teammate fragment.
 */
public class TeammateFragment extends ADataProgressFragment<ITeammateActivity> implements VoterBar.VoterBarListener {

    private static final String OBJECT_FRAGMENT_TAG = "object_tag";
    private static final String VOTING_TAG = "voting_tag";
    private static final String VOTING_STATS_FRAGMENT_TAG = "voting_stats_tag";


    private ImageView mUserPicture;
    private ImageView mSmallImagePicture;
    private ImageView mTeammateIcon;
    private TeambrellaAvatarsWidgets mAvatars;
    private NestedScrollView mScrollView;

    private TextView mUserName;
    private TextView mMessage;
    private TextView mUnread;
    private TextView mWhen;


    private View mCoverMeSection;
    private View mCoverThemSection;
    private View mWouldCoverPanel;
    private TextView mCoverMe;

    private TextView mCoverThem;
    private TextView mWouldCoverMe;
    private TextView mWouldCoverThem;
    private TextView mCoversMeTitle;
    private TextView mCoversThemTitle;


    private String mUserId;
    private int mTeamId;
    private String mTopicId;
    private String mCurrency;
    private int mTeamAccessLevel;

    private float mHeCoversMeIf02;
    private float mHeCoversMeIf1;
    private float mHeCoversMeIf499;
    private float mMyRisk;


    private boolean mIsShown;


    //"HeCoversMeIf02":10.104229912048861,"HeCoversMeIf1":45.50422458110436,"HeCoversMeIf499":46.31511561896891,"MyRisk":0.9

    @Override
    protected View onCreateContentView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_teammate, container, false);
        mUserPicture = view.findViewById(R.id.user_picture);
        mSmallImagePicture = view.findViewById(R.id.small_teammate_icon);
        mUserName = view.findViewById(R.id.user_name);
        mCoverMe = view.findViewById(R.id.cover_me);
        mCoverThem = view.findViewById(R.id.cover_them);
        mTeammateIcon = view.findViewById(R.id.teammate_icon);
        mAvatars = view.findViewById(R.id.avatars);
        mMessage = view.findViewById(R.id.message);
        mUnread = view.findViewById(R.id.unread);
        mCoverMeSection = view.findViewById(R.id.cover_me_section);
        mCoverThemSection = view.findViewById(R.id.cover_them_section);
        mWhen = view.findViewById(R.id.when);
        mWouldCoverPanel = view.findViewById(R.id.would_cover_panel);
        mScrollView = view.findViewById(R.id.scroll_view);
        mWouldCoverMe = view.findViewById(R.id.would_cover_me);
        mWouldCoverThem = view.findViewById(R.id.would_cover_them);
        mCoversMeTitle = view.findViewById(R.id.covers_me_title);
        mCoversThemTitle = view.findViewById(R.id.covers_them_title);
        if (savedInstanceState == null) {
            mDataHost.load(mTags[0]);
            setContentShown(false);
        }
        view.findViewById(R.id.discussion).setOnClickListener(v ->
                mDataHost.launchActivity(ChatActivity.getTeammateChat(getContext(), mTeamId, mUserId, null, null, mTopicId, mTeamAccessLevel)));
        return view;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        FragmentManager fragmentManager = getChildFragmentManager();

        FragmentTransaction transaction = fragmentManager.beginTransaction();
        if (fragmentManager.findFragmentByTag(OBJECT_FRAGMENT_TAG) == null) {
            transaction.add(R.id.object_info_container, ADataFragment.getInstance(mTags, TeammateObjectFragment.class), OBJECT_FRAGMENT_TAG);
        }

        if (fragmentManager.findFragmentByTag(VOTING_STATS_FRAGMENT_TAG) == null) {
            transaction.add(R.id.voting_statistics_container, ADataFragment.getInstance(mTags, TeammateVotingStatsFragment.class), VOTING_STATS_FRAGMENT_TAG);
        }


        if (fragmentManager.findFragmentByTag(VOTING_TAG) == null) {
            transaction.add(R.id.voting_container, ADataFragment.getInstance(mTags, TeammateVotingFragment.class), VOTING_TAG);
        }

        if (!transaction.isEmpty()) {
            transaction.commit();
        }
        mIsShown = false;
    }

    @Override
    protected void onDataUpdated(Notification<JsonObject> notification) {
        if (notification.isOnNext()) {
            Observable<JsonWrapper> responseObservable = Observable.just(notification.getValue())
                    .map(JsonWrapper::new);

            //noinspection unused
            final Integer matchId = responseObservable.map(jsonWrapper -> jsonWrapper.getObject(TeambrellaModel.ATTR_STATUS))
                    .map(jsonWrapper -> Uri.parse(jsonWrapper.getString(TeambrellaModel.ATTR_STATUS_URI)))
                    .map(TeambrellaUris.sUriMatcher::match)
                    .blockingFirst();


            Picasso picasso = TeambrellaImageLoader.getInstance(getContext()).getPicasso();


            Observable<JsonWrapper> dataObservable =
                    responseObservable.map(item -> item.getObject(TeambrellaModel.ATTR_DATA));

            Observable<JsonWrapper> basicObservable =
                    dataObservable.map(jsonWrapper -> jsonWrapper.getObject(TeambrellaModel.ATTR_DATA_ONE_BASIC));


            dataObservable.map(jsonWrapper -> jsonWrapper.getObject(TeambrellaModel.ATTR_DATA_ONE_TEAM))
                    .doOnNext(jsonWrapper -> mCurrency = jsonWrapper.getString(TeambrellaModel.ATTR_DATA_CURRENCY, mCurrency))
                    .doOnNext(jsonWrapper -> mTeamAccessLevel = jsonWrapper.getInt(TeambrellaModel.ATTR_DATA_TEAM_ACCESS_LEVEL, mTeamAccessLevel))
                    .onErrorReturnItem(new JsonWrapper(new JsonObject())).blockingFirst();


            basicObservable.doOnNext(basic -> AmountCurrencyUtil.setAmount(mCoverMe, basic.getFloat(TeambrellaModel.ATTR_DATA_COVER_ME), mCurrency))
                    .doOnNext(basic -> AmountCurrencyUtil.setAmount(mCoverThem, basic.getFloat(TeambrellaModel.ATTR_DATA_COVER_THEM), mCurrency))
                    .doOnNext(basic -> AmountCurrencyUtil.setAmount(mWouldCoverMe, basic.getFloat(TeambrellaModel.ATTR_DATA_COVER_ME), mCurrency))
                    .doOnNext(basic -> AmountCurrencyUtil.setAmount(mWouldCoverThem, basic.getFloat(TeambrellaModel.ATTR_DATA_COVER_THEM), mCurrency))
                    .doOnNext(basic -> mUserName.setText(basic.getString(TeambrellaModel.ATTR_DATA_NAME)))
                    .doOnNext(basic -> mTeamId = basic.getInt(TeambrellaModel.ATTR_DATA_TEAM_ID))
                    .doOnNext(basic -> mUserId = basic.getString(TeambrellaModel.ATTR_DATA_USER_ID))
                    .subscribe(jsonWrapper -> {
                    }, throwable -> {
                    }, () -> {
                    });


            basicObservable.map(basic -> TeambrellaServer.BASE_URL + basic.getString(TeambrellaModel.ATTR_DATA_AVATAR))
                    .doOnNext(uri -> picasso.load(uri).into(mUserPicture))
                    .doOnNext(uri -> picasso.load(uri).transform(new CropCircleTransformation()).into(mTeammateIcon))
                    .doOnNext(uri -> picasso.load(uri).into(mSmallImagePicture))
                    .onErrorReturnItem("").blockingFirst();


            dataObservable.map(data -> data.getObject(TeambrellaModel.ATTR_DATA_ONE_VOTING))
                    .doOnNext(voting -> {
                        View view = getView();
                        if (view != null) {
                            view.findViewById(R.id.voting_container).setVisibility(View.VISIBLE);
                            view.findViewById(R.id.object_info_container).setBackgroundResource(R.drawable.block);
                            mCoversMeTitle.setText(R.string.would_cover_me);
                            mCoversThemTitle.setText(R.string.would_cover_them);
                        }
                    })
                    .onErrorReturnItem(new JsonWrapper(null)).blockingFirst();

            Observable<JsonWrapper> discussionsObservable = dataObservable.map(data -> data.getObject(TeambrellaModel.ATTR_DATA_ONE_DISCUSSION));

            discussionsObservable.doOnNext(discussion -> mUnread.setText(discussion.getString(TeambrellaModel.ATTR_DATA_UNREAD_COUNT)))
                    .doOnNext(discussion -> mUnread.setVisibility(discussion.getInt(TeambrellaModel.ATTR_DATA_UNREAD_COUNT) > 0 ? View.VISIBLE : View.INVISIBLE))
                    .doOnNext(discussion -> mMessage.setText(Html.fromHtml(discussion.getString(TeambrellaModel.ATTR_DATA_ORIGINAL_POST_TEXT, null))))
                    .doOnNext(discussion -> mTopicId = discussion.getString(TeambrellaModel.ATTR_DATA_TOPIC_ID))
                    .doOnNext(discussion -> mWhen.setText(TeambrellaDateUtils.getRelativeTime(-discussion.getLong(TeambrellaModel.ATTR_DATA_SINCE_LAST_POST_MINUTES, 0))))
                    .onErrorReturnItem(new JsonWrapper(null)).blockingFirst();


            discussionsObservable.flatMap(discussion -> Observable.fromIterable(discussion.getJsonArray(TeambrellaModel.ATTR_DATA_TOP_POSTER_AVATARS)))
                    .map(jsonElement -> TeambrellaServer.BASE_URL + jsonElement.getAsString())
                    .toList()
                    .subscribe(mAvatars::setAvatars, e -> {
                    });

            mCoverThemSection.setVisibility(mDataHost.isItMe() ? View.GONE : View.VISIBLE);
            mCoverMeSection.setVisibility(mDataHost.isItMe() ? View.GONE : View.VISIBLE);

            dataObservable.map(data -> data.getObject(TeambrellaModel.ATTR_DATA_ONE_RISK_SCALE))
                    .doOnNext(riskScale -> {
                        mHeCoversMeIf1 = riskScale.getFloat(TeambrellaModel.ATTR_DATA_HE_COVERS_ME_IF1, mHeCoversMeIf1);
                        mHeCoversMeIf02 = riskScale.getFloat(TeambrellaModel.ATTR_DATA_HE_COVERS_ME02, mHeCoversMeIf02);
                        mHeCoversMeIf499 = riskScale.getFloat(TeambrellaModel.ATTR_DATA_HE_COVERS_ME_IF499, mHeCoversMeIf499);
                        mMyRisk = riskScale.getFloat(TeambrellaModel.ATTR_DATA_MY_RISK, mMyRisk);
                    }).onErrorReturnItem(new JsonWrapper(null)).blockingFirst();


            setContentShown(true);
            mIsShown = true;
        } else {
            setContentShown(true, !mIsShown);
            //noinspection ConstantConditions
            mDataHost.showSnackBar(ConnectivityUtils.isNetworkAvailable(getContext()) ? R.string.something_went_wrong_error : R.string.no_internet_connection);
        }
    }


    @Override
    public void onVoteChanged(float vote, boolean fromUser) {
        Rect scrollBounds = new Rect();
        mScrollView.getHitRect(scrollBounds);
        if (!mUserPicture.getLocalVisibleRect(scrollBounds) && mWouldCoverPanel.getVisibility() == View.GONE
                && fromUser) {
            ObjectAnimator animator = ObjectAnimator.ofFloat(mWouldCoverPanel, "translationY", -(float) mWouldCoverPanel.getHeight(), 0f).setDuration(300);
            animator.addListener(new AnimatorListenerAdapter() {
                @Override
                public void onAnimationStart(Animator animation) {
                    mWouldCoverPanel.setVisibility(View.VISIBLE);
                }
            });
            animator.start();
        }

        float value = (float) Math.pow(25, vote) / 5;

        if (value >= 0.2f && value <= 1f) {
            float wouldCoverMe = getLinearPoint(value, 0.2f, mHeCoversMeIf02, 1f, mHeCoversMeIf1);
            float wouldCoverThem = wouldCoverMe * mMyRisk / value;
            AmountCurrencyUtil.setAmount(mWouldCoverMe, wouldCoverMe, mCurrency);
            AmountCurrencyUtil.setAmount(mCoverMe, wouldCoverMe, mCurrency);
            AmountCurrencyUtil.setAmount(mWouldCoverThem, wouldCoverThem, mCurrency);
            AmountCurrencyUtil.setAmount(mCoverThem, wouldCoverThem, mCurrency);
        } else {
            float wouldCoverMe = getLinearPoint(value, 1f, mHeCoversMeIf1, 4.99f, mHeCoversMeIf499);
            float wouldCoverThem = wouldCoverMe * mMyRisk / value;
            AmountCurrencyUtil.setAmount(mWouldCoverMe, wouldCoverMe, mCurrency);
            AmountCurrencyUtil.setAmount(mCoverMe, wouldCoverMe, mCurrency);
            AmountCurrencyUtil.setAmount(mWouldCoverThem, wouldCoverThem, mCurrency);
            AmountCurrencyUtil.setAmount(mCoverThem, wouldCoverThem, mCurrency);
        }

        mWouldCoverPanel.removeCallbacks(mHideWouldCoverPanelRunnable);
    }

    @Override
    public void onVoterBarReleased(float vote, boolean fromUser) {
        if (mWouldCoverPanel.getVisibility() == View.VISIBLE) {
            mWouldCoverPanel.postDelayed(mHideWouldCoverPanelRunnable, 1000);
        }
    }


    private static float getLinearPoint(float x, float x1, float y1, float x2, float y2) {
        return ((x - x1) * (y2 - y1)) / (x2 - x1) + y1;
    }

    private Runnable mHideWouldCoverPanelRunnable = () -> {
        ObjectAnimator animator = ObjectAnimator.ofFloat(mWouldCoverPanel, "translationY", 0f, -(float) mWouldCoverPanel.getHeight()).setDuration(300);

        animator.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationEnd(Animator animation) {
                super.onAnimationEnd(animation);
                mWouldCoverPanel.setVisibility(View.GONE);
            }
        });

        animator.start();
    };
}

File: app/src/main/java/com/teambrella/android/ui/teammate/TeammateFragment.kt
package com.teambrella.android.ui.teammate

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.ObjectAnimator
import android.graphics.Rect
import android.os.Bundle
import android.support.v4.widget.NestedScrollView
import android.text.Html
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.google.gson.JsonObject
import com.teambrella.android.R
import com.teambrella.android.api.*
import com.teambrella.android.ui.base.ADataProgressFragment
import com.teambrella.android.ui.base.createDataFragment
import com.teambrella.android.ui.chat.ChatActivity
import com.teambrella.android.ui.util.setAvatar
import com.teambrella.android.ui.widget.TeambrellaAvatarsWidgets
import com.teambrella.android.ui.widget.VoterBar
import com.teambrella.android.util.AmountCurrencyUtil
import com.teambrella.android.util.ConnectivityUtils
import com.teambrella.android.util.TeambrellaDateUtils
import io.reactivex.Notification

class TeammateFragment : ADataProgressFragment<ITeammateActivity>(), VoterBar.VoterBarListener {

    private companion object {
        private const val OBJECT_FRAGMENT_TAG = "object_tag"
        private const val VOTING_TAG = "voting_tag"
        private const val VOTING_STATS_FRAGMENT_TAG = "voting_stats_tag"
        private const val VOTING_RESULT_TAG = "voting_result_tag"
        private const val CONTACTS_TAG = "contacts_tag"

        private fun getLinearPoint(x: Float, x1: Float, y1: Float, x2: Float, y2: Float): Float {
            return (x - x1) * (y2 - y1) / (x2 - x1) + y1
        }
    }


    private val userPictureView: ImageView? by ViewHolder(R.id.user_picture)
    private val smallImagePictureView: ImageView? by ViewHolder(R.id.small_teammate_icon)
    private val teammateIconView: ImageView? by ViewHolder(R.id.teammate_icon)
    private val avatarsView: TeambrellaAvatarsWidgets? by ViewHolder(R.id.avatars)
    private val scrollViewView: NestedScrollView? by ViewHolder(R.id.scroll_view)
    private val userNameView: TextView? by ViewHolder(R.id.user_name)
    private val messageView: TextView? by ViewHolder(R.id.message)
    private val unreadView: TextView? by ViewHolder(R.id.unread)
    private val whenView: TextView? by ViewHolder(R.id.`when`)
    private val coverMeSectionView: View? by ViewHolder(R.id.cover_me_section)
    private val coverThemSectionView: View? by ViewHolder(R.id.cover_them_section)
    private val wouldCoverPanelView: View? by ViewHolder(R.id.would_cover_panel)
    private val coverMeView: TextView? by ViewHolder(R.id.cover_me)
    private val coverThemView: TextView? by ViewHolder(R.id.cover_them)
    private val wouldCoverMeView: TextView? by ViewHolder(R.id.would_cover_me)
    private val wouldCoverThemView: TextView? by ViewHolder(R.id.would_cover_them)
    private val coversMeTitleView: TextView? by ViewHolder(R.id.covers_me_title)
    private val coversThemTitleView: TextView? by ViewHolder(R.id.covers_them_title)
    private val wouldCoverThemTitleView: TextView? by ViewHolder(R.id.would_cover_them_title)
    private val cityView: TextView? by ViewHolder(R.id.city)
    private val memberSinceView: TextView? by ViewHolder(R.id.member_since)
    private val votingContainerView: View? by ViewHolder(R.id.voting_container)
    private val objectInfoContainerView: View? by ViewHolder(R.id.object_info_container)
    private val votingResultContainerView: View? by ViewHolder(R.id.voting_result_container)


    private var currency: String? = null
    private var teamAccessLevel: Int = TeambrellaModel.TeamAccessLevel.FULL_ACCESS
    private var wouldCoverMeValue: Float? = null
    private var wouldCoverThemValue: Float? = null
    private var teamId: Int? = null
    private var userId: String? = null
    private var mGender: Int? = null
    private var topicId: String? = null

    private var heCoversMeIf02: Float = 0f
    private var heCoversMeIf1: Float = 0f
    private var heCoversMeIf499: Float = 0f
    private var myRisk: Float = 0f
    private var isShown = false


    private var mCoverageUpdateStarted: Boolean = false


    override fun onCreateContentView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?) =
            inflater.inflate(R.layout.fragment_teammate, container, false)
                    ?: throw RuntimeException()

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        childFragmentManager.let {
            val transaction = it.beginTransaction()

            if (it.findFragmentByTag(OBJECT_FRAGMENT_TAG) == null) {
                transaction.add(R.id.object_info_container
                        , createDataFragment(tags, TeammateObjectFragment::class.java)
                        , OBJECT_FRAGMENT_TAG)

                if (it.findFragmentByTag(VOTING_STATS_FRAGMENT_TAG) == null) {
                    transaction.add(R.id.voting_statistics_container
                            , getFragmentInstance(tags)
                            , VOTING_STATS_FRAGMENT_TAG)
                }


                if (it.findFragmentByTag(VOTING_TAG) == null) {
                    transaction.add(R.id.voting_container
                            , createDataFragment(tags, TeammateVotingFragment::class.java)
                            , VOTING_TAG)
                }

                if (it.findFragmentByTag(VOTING_RESULT_TAG) == null) {
                    transaction.add(R.id.voting_result_container
                            , createDataFragment(tags, TeammateVotingResultFragment::class.java)
                            , VOTING_RESULT_TAG)
                }

                if (it.findFragmentByTag(CONTACTS_TAG) == null) {
                    transaction.add(R.id.contacts_container
                            , createDataFragment(tags, TeammateContactsFragment::class.java)
                            , CONTACTS_TAG)
                }

            }

            if (!transaction.isEmpty) {
                transaction.commit()
            }
        }

        if (savedInstanceState == null) {
            dataHost.load(tags[0])
        }

        setContentShown(false)

        view.findViewById<View>(R.id.discussion_foreground).setOnClickListener {
            dataHost.launchActivity(ChatActivity.getTeammateChat(context, teamId
                    ?: 0, userId, null, null, topicId, teamAccessLevel))
        }

        isShown = false
        super.onViewCreated(view, savedInstanceState)
    }

    override fun onDataUpdated(notification: Notification<JsonObject>) {
        if (notification.isOnNext) {
            val data = notification.value?.data
            val basic = data?.basic

            data?.teamPart?.let {
                currency = it.currency ?: currency
                teamAccessLevel = it.teamAccessLevel ?: teamAccessLevel
            }

            basic?.let { _basic ->

                AmountCurrencyUtil.setAmount(coverMeView, wouldCoverMeValue
                        ?: _basic.coverMe ?: 0f, currency)

                AmountCurrencyUtil.setAmount(coverThemView, wouldCoverThemValue
                        ?: _basic.coverThem ?: 0f, currency)

                AmountCurrencyUtil.setAmount(wouldCoverMeView, wouldCoverMeValue
                        ?: _basic.coverMe ?: 0f, currency)

                AmountCurrencyUtil.setAmount(wouldCoverThemView, wouldCoverThemValue
                        ?: _basic.coverThem ?: 0f, currency)

                userNameView?.text = _basic.name

                teamId = _basic.teamId
                userId = _basic.userId
                mGender = _basic.gender ?: mGender

                cityView?.text = _basic.city

                _basic.dateJoined?.let {
                    memberSinceView?.text = getString(R.string.member_since_format_string
                            , TeambrellaDateUtils.getDatePresentation(context, TeambrellaDateUtils.TEAMBRELLA_UI_DATE, it))
                }


                val avatarString = _basic.avatar

                avatarString?.let {
                    val url = imageLoader.getImageUrl(it)
                    userPictureView?.setAvatar(url)
                    smallImagePictureView?.setAvatar(url)
                    teammateIconView?.setAvatar(url)
                }

            }

            val voting = data?.voting

            if (voting != null) {
                votingContainerView?.visibility = View.VISIBLE
                objectInfoContainerView?.setBackgroundResource(R.drawable.block)
                coversMeTitleView?.setText(R.string.would_cover_me)

                when (mGender) {
                    TeambrellaModel.Gender.MALE -> {
                        coversThemTitleView?.setText(R.string.would_cover_him)
                        wouldCoverThemTitleView?.setText(R.string.would_cover_him)
                    }
                    TeambrellaModel.Gender.FEMALE -> {
                        coversThemTitleView?.setText(R.string.would_cover_her)
                        wouldCoverThemTitleView?.setText(R.string.would_cover_her)
                    }
                    else -> {
                        coversThemTitleView?.setText(R.string.would_cover_them)
                        wouldCoverThemTitleView?.setText(R.string.would_cover_them)
                    }
                }
            } else {
                when (mGender) {
                    TeambrellaModel.Gender.MALE -> {
                        coversThemTitleView?.setText(R.string.cover_him)
                    }
                    TeambrellaModel.Gender.FEMALE -> {
                        coversThemTitleView?.setText(R.string.cover_her)
                    }
                    else -> {
                        coversThemTitleView?.setText(R.string.cover_them)
                    }
                }
            }

            data?.voted?.let {
                votingResultContainerView?.visibility = View.VISIBLE
            }

            data?.discussionPart?.let { _discussionPart ->
                unreadView?.text = _discussionPart.unreadCount?.toString()
                unreadView?.visibility = if (_discussionPart.unreadCount ?: 0 > 0) View.VISIBLE else View.GONE

                _discussionPart.originalPostText?.let { _text ->
                    @Suppress("DEPRECATION")
                    messageView?.text = Html.fromHtml(_text)
                }

                topicId = _discussionPart.topicId ?: topicId
                whenView?.text = TeambrellaDateUtils.getRelativeTime(-(_discussionPart.sinceLastPostMinutes
                        ?: 0L))

                val posterCount = _discussionPart.posterCount ?: 0
                _discussionPart.topPosterAvatars?.map { it.asString }?.let {
                    avatarsView?.setAvatars(imageLoader, it, posterCount)
                }

            }

            coverThemSectionView?.visibility = if (dataHost.isItMe) View.GONE else View.VISIBLE
            coverMeSectionView?.visibility = if (dataHost.isItMe) View.GONE else View.VISIBLE

            data?.riskScale?.let {
                heCoversMeIf1 = it.heCoversMeIf1 ?: heCoversMeIf1
                heCoversMeIf02 = it.hetCoversMeIf02 ?: heCoversMeIf02
                heCoversMeIf499 = it.heCoversMeIf499 ?: heCoversMeIf499
                myRisk = it.myRisk ?: myRisk
            }

            isShown = true
            setContentShown(true)
        } else {
            setContentShown(true, !isShown)
            dataHost.showSnackBar(if (ConnectivityUtils.isNetworkAvailable(context)) R.string.something_went_wrong_error else R.string.no_internet_connection)
        }
    }


    override fun onVoteChanged(vote: Float, fromUser: Boolean) {
        val scrollBounds = Rect()
        scrollViewView?.getHitRect(scrollBounds)
        if (userPictureView?.getLocalVisibleRect(scrollBounds) == false && wouldCoverPanelView?.visibility == View.INVISIBLE
                && fromUser && !dataHost.isItMe) {
            val animator = ObjectAnimator.ofFloat(wouldCoverPanelView, "translationY", -(wouldCoverPanelView?.height?.toFloat()
                    ?: 0f), 0f).setDuration(300)
            animator.addListener(object : AnimatorListenerAdapter() {
                override fun onAnimationStart(animation: Animator) {
                    wouldCoverPanelView?.visibility = View.VISIBLE
                }
            })
            animator.start()
        }

        val value = Math.pow(25.0, vote.toDouble()).toFloat() / 5

        if (value in 0.2f..1f) {
            wouldCoverMeValue = getLinearPoint(value, 0.2f, heCoversMeIf02, 1f, heCoversMeIf1)
            wouldCoverThemValue = wouldCoverMeValue!! * myRisk / value
        } else {
            wouldCoverMeValue = getLinearPoint(value, 1f, heCoversMeIf1, 4.99f, heCoversMeIf499)
            wouldCoverThemValue = wouldCoverMeValue!! * myRisk / value
        }

        if (!mCoverageUpdateStarted) {
            wouldCoverPanelView?.post(mCoverageUpdate)
            mCoverageUpdateStarted = true
        }

        wouldCoverPanelView?.removeCallbacks(mHideWouldCoverPanelRunnable)
    }

    override fun onVoterBarReleased(vote: Float, fromUser: Boolean) {
        if (wouldCoverPanelView?.visibility == View.VISIBLE) {
            wouldCoverPanelView?.postDelayed(mHideWouldCoverPanelRunnable, 1000)
        }
        wouldCoverPanelView?.removeCallbacks(mCoverageUpdate)
        mCoverageUpdateStarted = false
    }


    private val mHideWouldCoverPanelRunnable = {
        val animator = ObjectAnimator.ofFloat(wouldCoverPanelView, "translationY", 0f, -(wouldCoverPanelView?.height?.toFloat()
                ?: 0f)).setDuration(300)

        animator.addListener(object : AnimatorListenerAdapter() {
            override fun onAnimationEnd(animation: Animator) {
                super.onAnimationEnd(animation)
                wouldCoverPanelView?.visibility = View.INVISIBLE
            }
        })

        animator.start()
    }

    private val mCoverageUpdate = object : Runnable {
        override fun run() {
            if (wouldCoverPanelView != null) {
                AmountCurrencyUtil.setAmount(wouldCoverMeView, wouldCoverMeValue ?: 0f, currency)
                AmountCurrencyUtil.setAmount(coverMeView, wouldCoverMeValue ?: 0f, currency)
                AmountCurrencyUtil.setAmount(wouldCoverThemView, wouldCoverThemValue
                        ?: 0f, currency)
                AmountCurrencyUtil.setAmount(coverThemView, wouldCoverThemValue ?: 0f, currency)
                wouldCoverPanelView?.postDelayed(this, 100)
            }

        }
    }
}